src/mscorlib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5994,7 +5994,10 @@ public static Task WhenAll(IEnumerable<Task> tasks)
             foreach (Task task in tasks)
             {
                 if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
-                taskList.Add(task);
+                if (!task.IsRanToCompletion) // no need to wait on the task if it's already successfully completed
```

Comment:
It looks like previously it would have transferred wait notifications from completed tasks to the when-all task here:
https://github.com/dotnet/coreclr/pull/8567/files#diff-9e9a920444355b78dc1fb28c79fec014R6107
and here:
https://github.com/dotnet/coreclr/pull/8567/files#diff-9e9a920444355b78dc1fb28c79fec014R6142

Does that still need to happen?

LGTM otherwise